aip/general/0129.md

```diff
@@ -0,0 +1,150 @@
+---
+id: 129
+state: draft
+created: 2023-06-21
+placement:
+  category: resource-design
+  order: 65
+---
+
+# Server-Modified Values and Defaults
+
+Server-modified and default values often result in a poor user experience for
+[state-driven clients](state-driven-clients). These clients are often unable to
+tell when their desired state matches the current state for these fields, as the
+rules by which a server may modify and return values are complex, not public,
+and not repeatable. This AIP aims to clarify what behaviors can be expected from
+services and describing how they can be repeated on the client-side.
+
+## Guidance
+
+### Single Owner Fields
+
+Fields should have a single owner, whether that is the client or the server.
+Server owned fields should be indicated with the `OUTPUT_ONLY` field_behavior.
+All other types of fields should be considered to be owned by the client. The
+server should respect the value (or lack thereof) for all client owned fields
+and not modify them. Failure to respect client set values can lead to
+[Declarative clients][] entering an infinite loop to correct the change.
+
+### Effective Values
+
+There are instances where a service will allocate, generate, or calculate a
+value if the client chooses not to specify one. For example: a client may
+create a virtual machine without specifying a static IP address for the virtual
+machine to be available on. Such a scenario is opting into dynamic IP address
+allocation.
+
+Some examples of these types of fields are ones that are:
+
+* generated (UUID)
+* allocated (dynamic IP address)
+* assigned (most recent software package version)
+
+An attribute with an effective value ***must*** be expressed as two fields in
+the API:
+* a mutable field that can be optionally set by the user and **must not** be
+modified by the service
+* an `OUTPUT_ONLY` field that records the effective value decided on by the
+service
+
+Example:
+```proto
+message VirtualMachine {
+  …
+  string ip_address = 4;
+  string effective_ip_address = 5 [
+    (google.api.field_behavior) = OUTPUT_ONLY
+  ];
+}
+```
+
+#### Naming
+
+Effective values **must** be named by prefixing `effective_` to the mutable
+field's name.
+
+### User-Specified Fields
+
+For user-specified fields, the value in response from the service **must** be
+the same as provided by the create or update request. For string fields this
+means returning the value unchanged, with two exceptions: 
+
+* When a string field has a data type annotation, a normalized string that
+represents the given value may be returned.
+* When a string field has a default value indicated by the `default` keyword,
```

Comment:
Apologies if this comment is off-base, as I'm not familiar with the constraints of protobufs, but per my comment above regarding dynamic vs. defaulted-static IP address allocation, there are scenarios where the value has a true default but that value is not fixed.  Phrased differently: there seems an important distinction between a server-calculated default for a property that remains client-managed, and a default that causes the property to become system-managed (such as a DHCP-managed IP address).

Reply:
Is that distinction important?

From my side it's important to determine what fields can be set by a state-driven client and what fields we expect the server to communicate information to the client via server-set fields. Those should not mix. Both a server-calculated default IP address *or* a dynamic IP address can be examples of server-set values that can be communicated via output-only fields

Reply:
To my view, the key distinction is whether the property is _managed_ by the client or by the server.  A property  that has its value set at resource creation but will never be subsequently changed by the server remains client-managed.  This is the case for the exception carved out above for `default`, but is also true for more complex scenarios that cannot be expressed through a `default` clause but nonetheless represent a one-time initialization of a property that remains client-managed.